### ISSUE TITLE
What's the difference between `uv pip install` and `uv add`

### ISSUE THREAD

**Issue by lucharo** [astral-sh/uv#9219]
I've been using `uv` for a while and I really enjoy it. I keep stumbling onto the same confusion though. I never quite know whether do to:

```sh
uv init 
uv venv 
uv add polars marimo
uv run hello.py
```

or 

```sh
uv init 
uv venv 
source .venv/bin/activate
pip install polars marimo
python hello.py
```

are these two above equivalent? 

---

also are these two equivalent?
```
uv add polars 
```

```
uv pip install polars
```

**Comment by charliermarsh** [comment]
They are two separate APIs for managing your Python project and environment.

The `uv pip` APIs are meant to resemble the pip CLI. You can think of this as a slightly "lower-level" API: you tell `uv pip` to install a specific package, or remove a specific package, and so on.

`uv add`, `uv run`, `uv sync`, and `uv lock` are what we call the "project APIs". These are "higher-level": you define your dependencies in `pyproject.toml`, and uv ensures that your environment is always in-sync with those dependencies.

The project APIs are more opinionated (you _must_ use `pyproject.toml`, since they're designed around "projects"), while the `uv pip` APIs are more flexible (you can manipulate a virtual environment however you want -- there aren't really any "rules").

If you're starting a new project, we recommend using the project APIs. If you're working with existing projects, it's often easier to use the `uv pip` APIs, if those projects already have established workflows based on pip (since you can just replace `pip install` with `uv pip install`, etc.).

**Comment by charliermarsh** [comment]
So concretely:

```
uv init 
uv venv 
uv add polars marimo
uv run hello.py
```

This makes a lot of sense (although you can omit `uv venv` -- we'll create it automatically). You're creating a new project, adding some dependencies, then running a command with those dependencies installed.

```
uv init 
uv venv 
source .venv/bin/activate
pip install polars marimo
python hello.py
```

This is also reasonable, though you could omit `uv init` -- there's no need to make a project, since you're not taking advantage of the "project" capabilities (you're not adding `polars` and `marimo` to the project itself -- just to the virtual environment).

**Comment by charliermarsh** [comment]
I hope that's helpful. Let me know if you have any follow-up questions.

**Comment by rxy1212** [comment]
@charliermarsh When I use 'uv add' how to set another source like 'pip install -i' do?

**Comment by FishAlchemist** [comment]
@rxy1212  Actually, ``uv add`` has related flags, but it's better to write the source in pyproject.toml. Otherwise, when you update and lock lockfile next time, you'll need to bring in the related flags again.
```
      --index <INDEX>                        The URLs to use when resolving dependencies, in addition to the default
                                             index [env: UV_INDEX=]
      --default-index <DEFAULT_INDEX>        The URL of the default package index (by default:
                                             <https://pypi.org/simple>) [env: UV_DEFAULT_INDEX=]
  -i, --index-url <INDEX_URL>                (Deprecated: use `--default-index` instead) The URL of the Python package
                                             index (by default: <https://pypi.org/simple>) [env: UV_INDEX_URL=]
      --extra-index-url <EXTRA_INDEX_URL>    (Deprecated: use `--index` instead) Extra URLs of package indexes to use,
                                             in addition to `--index-url` [env: UV_EXTRA_INDEX_URL=]
```
``--index`` and ``--default-index``
https://docs.astral.sh/uv/configuration/indexes/

**Comment by rxy1212** [comment]
@FishAlchemist That's what I need. Thank you!

**Comment by charliermarsh** [comment]
Thanks @FishAlchemist, you're super helpful!

**Comment by lucharo** [comment]
Hey @charliermarsh, thank you so much for the clarifications. I do have some other questions. 

I am confused about uv and virtual  environment interact. what would happen in the following scenarios?

```bash
uv venv
uv add polars
## or
uv venv
uv pip install polars 
```

Do these install in the venv directly or do I need to `source  .venv/bin/activate`?

Using pip "traditionally", I would check `which pip` to see where I'd be installing my package but in `uv` I cannot really do that or can I? 


Also if I go for the `uv` API fully, I would never really run `python script.py` correct? It'd always be `uv run script.py`?

**Comment by zanieb** [comment]
Yes `uv pip install polars` will modify the `.venv` in the working directory unless you've activated another one. You can use `uv python find` to see which Python environment will be modified by a given `uv pip` operation.

**Comment by dmi3kno** [comment]
I know this issue is closed, but I am still slightly confused about the need to (re-)activate a project environment after i add new packages. Lets say i set up my environment like this

```{python}
uv init
uv add polars
uv run my_script.py
```
But then I realized I need a plotting library, so I changed my script and now I want to add it to dependencies

```{python}
uv init
uv add plotnine
# do i need to source .venv/bin/activate here?
uv run my_script.py
```
Do I need to re-activate the environment for the libraries to become available to my script? It seems like I did not have to do it the first time. 

In other words, when if ever, would one need to `source .venv/bin/activate` if one sticks to project API (i.e. modifies environment only through `uv add`, `uv remove`, `uv sync` and such)?

**Comment by FishAlchemist** [comment]
@dmi3kno When using ``uv run``, you don't need to explicitly enter the virtual environment. It will automatically locate the correct one. But, are you asking because you've encountered a problem?

**Note:** If you're running scripts with inline script metadata, uv will create a temporary virtual environment to execute them.
https://packaging.python.org/en/latest/specifications/inline-script-metadata/
https://docs.astral.sh/uv/guides/scripts/#declaring-script-dependencies

**Comment by akupiectdc** [comment]
> I hope that's helpful. Let me know if you have any follow-up questions.

I do, but they might sound silly. 
1. In the second option - is there a difference between uv pip install and pip install?
2. Also in the second option - what is the difference between uv run script.py and python script.py?

**Comment by jingyugao** [comment]
what is the different of `uv pip` and `uv run pip`? 
```
➜  ~ uv run pip --version
pip 24.3.1 from /Users/dev/.local/share/uv/python/cpython-3.12.9-macos-aarch64-none/lib/python3.12/site-packages/pip (python 3.12)
➜  ~ uv pip --version
uv-pip 0.6.12 (Homebrew 2025-04-02)
```

**Comment by lmBored** [comment]
> what is the different of `uv pip` and `uv run pip`?
> 
> ```
> ➜  ~ uv run pip --version
> pip 24.3.1 from /Users/dev/.local/share/uv/python/cpython-3.12.9-macos-aarch64-none/lib/python3.12/site-packages/pip (python 3.12)
> ➜  ~ uv pip --version
> uv-pip 0.6.12 (Homebrew 2025-04-02)
> ```

`uv-pip` is a package from Homebrew, it's not `uv pip`. If you do brew uninstall uv-pip, then `uv run pip -V` again, you should get the same pip version

**Comment by ishaanjain** [comment]
> So concretely:
> 
> ```
> uv init 
> uv venv 
> uv add polars marimo
> uv run hello.py
> ```
> 
> This makes a lot of sense (although you can omit `uv venv` -- we'll create it automatically). You're creating a new project, adding some dependencies, then running a command with those dependencies installed.
> 
> ```
> uv init 
> uv venv 
> source .venv/bin/activate
> pip install polars marimo
> python hello.py
> ```
> 
> This is also reasonable, though you could omit `uv init` -- there's no need to make a project, since you're not taking advantage of the "project" capabilities (you're not adding `polars` and `marimo` to the project itself -- just to the virtual environment).

@charliermarsh I believe the `pip install polars marimo` command would not work, right? Because `uv venv` does not include pip by default.

**Comment by gigberg** [comment]
> uv init 
> uv venv 
> source .venv/bin/activate
> pip install polars marimo
> python hello.py

**Updated at  2025-09-27 !**
Because `uv venv` not install `python -m pip`(traditional pip command) itself, and subcommand `uv pip` is not equal (but functionally similar) to `python -m pip`. So their is two workflow (or interface) in uv:
1.First, for pure uv user, you are supposed to use uv project workflow: 
1)`uv init` to generate scaffold  files, like `.pyproject.toml` and `.git`,
~~2)`uv venv --python 3.12` to generate `.venv` folder, (like other python virtualenv tools do), (can be ignored)~~
3)`uv add xxx` to modify  `.pyproject.toml` in order to install new python packages (dependicies) to this `.venv`.
4)`uv run xxx.py` to execute python scripts with (in) this `.venv` (not `uv python xxx.py` which is illegal usage)

*`uv venv` can be ignored since `uv add` would trigger this automaticly*
**the uv project workflow never activate `.venv` explicitly (so does not interfere with system $PATH). Moreover, the uv project workflow has nothing to do with `python -m pip` or `uv pip`; it uses its own built-in `dependency resolution`, `package downloading`, and `package installation` functionality."**

2.Second, for traditional `python -m pip` command user, a python-pip-compatible workflow called `uv pip` is provided (Unlike the project concept and `.pyproject` uv.lock file which are designed specifically for making PyPI packages,  `uv pip`  is more suitable for instant use.):
1)`uv venv --python 3.12` to generate `.venv` folder, 
2)`uv pip install xxx` to  install new python packages to this `.venv`. (similar function  to `python pip` but different internals)
3)`source activate .venv/bin/activate` to activate the `.venv` and modify system $PATH
4)`python xxx.py` to execute python scripts with (in) this `.venv`

**The uv pip interface is designed to help users who are accustomed to `python virtualenv` and `python pip` transition smoothly to the uv tool. It still requires creating, activating a virtual environment and modifies the $PATH environment variable as these old tools does. 
The key difference is that it uses uv's own `uv pip` subcommand instead of the standard `python pip` (and similarly, use `uv venv` subcommand instead of `python virtualenv`). The advantage of `uv pip` is that it leverages uv’s internal, more efficient `dependency resolution`, `package downloading`, and `package installation` mechanisms**.

------------------------original reply----------------------------------------
yes, `pip install polars marimo` will call pip in system-wide, if you are in ubuntu22.04, it's `/usr/bin/pip`, and therefore `pip install polars marimo` will try to install packages to system-wide path (or fallbacked to user-wide path since pip is not run with sudo)!! which is not in current directory's .venv dir as your espect. 

And since u are not run with sudo privilege, the error message will raise "**Defaulting to user installation because normal site-packages is not writeable**", which means `pip install polars marimo`  fallbacked to `pip install --user polars marimo`, which install packages to `/home/your_user_name/.local/lib/python3.10/site-packages/`. 

By the way, ubuntu default `pip install` in global and user site-packages path are isolated and can have same packages, so when  ubuntu default `python xxx.py` will search both site-packages directory and select the first find, usually user package prior to global package.

**Comment by ishaanjain** [comment]
Makes sense. So it seems
```
uv init
uv venv
source .venv/bin/activate
pip install polars marimo
python hello.py
```
is not really a sensible workflow. Better would be
```
uv venv
uv pip install polars marimo
source .venv/bin/activate
python hello.py
```

**Comment by lmBored** [comment]
> Makes sense. So it seems
> 
> ```
> uv init
> uv venv
> source .venv/bin/activate
> pip install polars marimo
> python hello.py
> ```
> 
> is not really a sensible workflow. Better would be
> 
> ```
> uv venv
> uv pip install polars marimo
> source .venv/bin/activate
> python hello.py
> ```

It was recommended above to use `uv add` instead of `uv pip install` since `uv add` manages lockfile and allow you to use `uv sync` and `uv lock -U` to sync dependencies between laptops

**Comment by robotrapta** [comment]
I found this discussion super elucidating.  It explains a lot of why `uv` sometimes "just works" and sometimes gets confusing and rat-hole-ish.

I think the main-page docs would be well-served to be more opinionated and say "We recommend you use projects, and then these are probably the only commands you'll need: `uv add`, `uv run`, `uv sync`, and `uv lock`."
Then under the `pip` section say "We recommend letting `uv` manage everything for you with the project commands, but if you know what you're doing, and need more control, here's a lower layer of commands..."

**Comment by dmi3kno** [comment]
> I think the main-page docs would be well-served to be more opinionated and say

This is actually a very good suggestion. I support 💯

**Comment by ishaanjain** [comment]
I agree the docs could be more opinionated. I wrote a brief blog post sharing my understanding of when to use `uv add` vs `uv pip install`, in case it's useful to others:
https://ishaanjain.hashnode.dev/uv
Happy to adjust if I’ve misunderstood anything.
